### PR TITLE
🏗  ♻️ Fix type errors in /build-system/tasks

### PR DESCRIPTION
**Pull request by rileyajones**
The ongoing effort to add typing to the build-system directory has finally reached the largest directory "tasks". Due to the large size of this directory the e2e, performance, and visual-diff subdirectories have been excluded from this PR but will be fixed in another upcoming PR.

The total number of type errors within the /build-system/tasks directory was 544 which combined with the additional 6 outstanding summed to 550.
After this PR there will be a total of 281 and without the e2e, performance, and visual-diff directories the number drops to 15.

#28387

### CODE REVIEW

File: build-system/tasks/sweep-experiments/index.js
Comment (rileyajones) [inline build-system/tasks/sweep-experiments/index.js:244]: Unlike my the other comment I left about string.split.shift this could technically be undefined but I trust that the @alanorozco understands this code better than I.

File: build-system/test-configs/dep-check-config.js
Comment (rileyajones) [inline build-system/test-configs/dep-check-config.js:34]: Moving to /build-system/tasks/dep-check.js

File: build-system/tsconfig.json
Comment (rileyajones) [inline build-system/tsconfig.json:30]: This should be `"./tasks/**/*.js",`

File: build-system/tsconfig.json
Comment (rileyajones) [inline build-system/tsconfig.json:63]: This change is a more accurate representation of the state of the type checking effort but eventually it will be be reverted to the more general includes list on previously here. I am fine with not updating this at all if others dislike the churn.

File: build-system/tasks/performance/helpers.js
Comment (danielrozenberg) [reply build-system/tasks/performance/helpers.js:175]: Docs say `If no encoding is specified (using options.encoding), the data is returned as a <Buffer> object. Otherwise, the data will be a string.` - since you need to make a change here, I think it's better to set an encoding

File: build-system/tasks/performance/rewrite-analytics-tags.js
Comment (danielrozenberg) [reply build-system/tasks/performance/rewrite-analytics-tags.js:49]: How about:
```js
const vendor = tag.getAttribute('type');
if (vendor) {
  ...
}
```

File: build-system/test-configs/karma.conf.js
Comment (danielrozenberg) [inline build-system/test-configs/karma.conf.js:32]: Feels kinda "off" to push a value onto a const (alas JS doesn't have true const arrays) - how about keeping the ?: syntax and just replace it null with `''`?

File: build-system/tasks/serve.js
Comment (danielrozenberg) [inline build-system/tasks/serve.js:132]: The readability here isn't 100% - how about `for (const detail of (device[1] ?? []))`?

File: build-system/tasks/runtime-test/helpers.js
Comment (danielrozenberg) [inline build-system/tasks/runtime-test/helpers.js:191]: This noop doesn't feel like a noop because it only functions if you consider the cross-file changes... I'd split this change into a separate PR

File: build-system/tasks/performance/helpers.js
Comment (danielrozenberg) [reply build-system/tasks/performance/helpers.js:175]: Also in build-system/tasks/performance/print-report.js

File: build-system/tasks/get-zindex/index.js
Comment (danielrozenberg) [inline build-system/tasks/get-zindex/index.js:63]: Could also just merge these two directly in the const: `const selectorNames = rule.selector.replace('\n', '').split(',');`

File: build-system/tasks/extension-generator/index.js
Comment (danielrozenberg) [inline build-system/tasks/extension-generator/index.js:316]: Instead of adding empty strings to every `execOrThrow` call, can you change the signature of that function to have a default value of `''` for the second arg?

File: build-system/pr-check/npm-checks.js
Comment (danielrozenberg) [inline build-system/pr-check/npm-checks.js:104]: It's unclear how this is a noop

File: build-system/tsconfig.json
Comment (estherkim) [inline build-system/tsconfig.json:31]: Can we simplify this to `./tasks/*` ? Looks like the only supported extensions are `.ts, .tsx, .d.ts, .js, .jsx` anyways.
https://www.typescriptlang.org/tsconfig#include

File: build-system/compile/compile.js
Comment (estherkim) [inline build-system/compile/compile.js:53]: `compile` takes `string[]|string` - is that accurate or should it only take `string[]`?

also nit: make param plural `entryModuleFilenames`

Comment (rsimha) [reply build-system/compile/compile.js:53]: Array or string is correct. (It's a relic of how we do things differently during dep checking. I'm writing in a reactor of this file, so feel free to ignore the arg renaming for now.)

File: build-system/tasks/extension-generator/index.js
Comment (rileyajones) [reply build-system/tasks/extension-generator/index.js]: There are only a couple places where the second argument is missing and enforcing better error messages seems like a good thing.

File: build-system/pr-check/npm-checks.js
Comment (rileyajones) [reply build-system/pr-check/npm-checks.js]: Neither of those functions accepts any arguments.

File: build-system/tsconfig.json
Comment (rileyajones) [reply build-system/tsconfig.json:31]: That is an excellent point.

File: build-system/compile/compile.js
Comment (rileyajones) [reply build-system/compile/compile.js:53]: `string[]|string` makes sense. There are no places where it is called with `string` so I guess I missed that it was an option.

File: build-system/compile/compile.js
Comment (rsimha) [reply build-system/compile/compile.js:153]: Good point. Let's move this to line 76. It's the top level that takes `timeInfo`.

File: build-system/pr-check/npm-checks.js
Comment (rsimha) [inline build-system/pr-check/npm-checks.js:102]: I love that this is catching bugs like this.

File: build-system/pr-check/utils.js
Comment (rsimha) [inline build-system/pr-check/utils.js:138]: We should just remove the `@return` annotation here and everywhere else that's now a `void`, right?

File: build-system/tasks/check-owners.js
Comment (rsimha) [reply build-system/tasks/check-owners.js:47]: Agree, passing in an array of globs is the intent.

File: build-system/tasks/check-video-interface-list.js
Comment (rsimha) [reply build-system/tasks/check-video-interface-list.js:52]: This works.

File: build-system/tasks/coverage-map/index.js
Comment (rsimha) [reply build-system/tasks/coverage-map/index.js:87]: It seems wrong to call `resolve` with `undefined`. Any way this can be fixed?

File: build-system/tasks/extension-generator/index.js
Comment (rsimha) [inline build-system/tasks/extension-generator/index.js:318]: Might as well make this message something like `Could not format files`

File: build-system/tasks/extension-generator/index.js
Comment (rsimha) [inline build-system/tasks/extension-generator/index.js:338]: This looks like it's introducing a silent failure. I think it should be accompanied by a warning.
```suggestion
  if (!bundleConfig) {
    log(yellow('WARNING:'), 'Could not write extension files.');
    return;
  }
```

File: build-system/tasks/extension-generator/insert-extension-bundles-config.js
Comment (rsimha) [inline build-system/tasks/extension-generator/insert-extension-bundles-config.js:61]: Same as before, make this an explicit message like `Could not format extension bundle`.

File: build-system/tasks/get-zindex/index.js
Comment (rsimha) [inline build-system/tasks/get-zindex/index.js:33]: nit: Insert a blank line between this block and the lines above. (Helps our linter to auto-sort `require` statements.)

File: build-system/tasks/get-zindex/index.js
Comment (rsimha) [reply build-system/tasks/get-zindex/index.js:63]: I think this is preferable to moving everything into the `if` block like you've done.

File: build-system/tasks/performance/helpers.js
Comment (rsimha) [reply build-system/tasks/performance/helpers.js:175]: Yep, best to just pass in an encoding. Also, agree that we should use the async version. 

```suggestion
  return await fs.readFile(filePath, `utf-8`);
```

File: build-system/tasks/report-test-status.js
Comment (rsimha) [inline build-system/tasks/report-test-status.js:142]: The intent was for `to GitHub` not to be colored.

```suggestion
      log('Reported', cyan(`${type}/${action}`), 'to GitHub');
```

File: build-system/tasks/runtime-test/helpers.js
Comment (rsimha) [reply build-system/tasks/runtime-test/helpers.js:191]: I recently wrote this code and can confirm that `browsers` is not used in any way, so okay with this change here. (The sentiment of Daniel's comment is worth following in other non-obvious cases though.)

File: build-system/tasks/runtime-test/runtime-test-base.js
Comment (rsimha) [inline build-system/tasks/runtime-test/runtime-test-base.js:385]: Most of these types aren't giving me any extra info. Can we ignore this one?

File: build-system/tasks/test-report-upload.js
Comment (rsimha) [inline build-system/tasks/test-report-upload.js:49]:
```suggestion
      (await fs.readFile(`result-reports/${testType}.json`, 'utf-8'))
```

File: build-system/test-configs/karma.conf.js
Comment (rsimha) [inline build-system/test-configs/karma.conf.js:30]: This inserts a blank string entry to the array while `.filter(Boolean)` removes it. Any other way to fix this?

**Review by rsimha — COMMENTED** [review]
Lots of excellent changes in this PR! Some comments below.

File: build-system/test-configs/karma.conf.js
Comment (rileyajones) [reply build-system/test-configs/karma.conf.js:30]: My previous solution prevented this. I preferred that approach but @danielrozenberg didn't like the idea of pushing to a `const` list.
`
if (argv.debug) {
  COMMON_CHROME_FLAGS.push('--auto-open-devtools-for-tabs');
}
`

Perhaps we could build the list in a function?

`
function getCommongChromeFlags() {
  let flags = [
    // Dramatically speeds up iframe creation time.
    '--disable-extensions',
    // Allows simulating user actions (e.g unmute) which otherwise will be denied.
    '--autoplay-policy=no-user-gesture-required',
  ];
  if (argv.debug) {
    flags.push('--auto-open-devtools-for-tabs');
  }
  return flags;
}

const COMMON_CHROME_FLAGS = getCommongChromeFlags();
`

File: build-system/tasks/runtime-test/runtime-test-base.js
Comment (rileyajones) [reply build-system/tasks/runtime-test/runtime-test-base.js]: Yeah, that makes sense. TS really doesn't like Object.assign

File: build-system/pr-check/utils.js
Comment (rileyajones) [reply build-system/pr-check/utils.js:138]: That should be fine.

File: build-system/test-configs/karma.conf.js
Comment (rsimha) [reply build-system/test-configs/karma.conf.js:30]: `karma.conf.js` is meant to be a collection of mostly statically determined initial config, so I don't think we should add a builder function to this file. (The more complicated dynamic config building is done by the `RuntimeTestConfig` class in `runtime-test-base.js`.)

To keep this simple, let's use the first option you listed here. (Here's a [discussion](https://stackoverflow.com/questions/23436437/why-can-i-change-value-of-a-constant-in-javascript) on pushing a value to a `const`.)

Comment (rileyajones) [reply build-system/test-configs/karma.conf.js:30]: Thanks for the discussion link,  that's always been my perspective on const Objects in JS but it's good to hear it reiterated.

File: build-system/pr-check/npm-checks.js
Comment (rsimha) [reply build-system/pr-check/npm-checks.js]: Yep, these args should go away. The fact that they're here is my fault 😃

File: build-system/tasks/coverage-map/index.js
Comment (rileyajones) [reply build-system/tasks/coverage-map/index.js]: I'll add a type assertion.

File: build-system/tasks/report-test-status.js
Comment (rsimha) [reply build-system/tasks/report-test-status.js:121]: Yeah, we're implicitly assuming that this code is only invoked by `e2e`, `integration`, and `unit`. Okay with this change here for now with a functional fix whenever this file needs to be updated next.

File: build-system/tasks/visual-diff/index.js
Comment (rsimha) [inline build-system/tasks/visual-diff/index.js:836]: These functions were deleted recently, so this might be the result of a bad merge. I'd ask @danielrozenberg for a careful review of this file.

**Review by rsimha — COMMENTED** [review]
Yay! LGTM from me after one important change to the visual diff file (bad merge).

Merge away after that's fixed and Daniel approves.

File: build-system/tasks/visual-diff/index.js
Comment (rileyajones) [reply build-system/tasks/visual-diff/index.js:836]: Great catch!

Comment (danielrozenberg) [reply build-system/tasks/visual-diff/index.js]: Is this file no longer part of this PR?

File: build-system/tasks/report-test-status.js
Comment (rileyajones) [reply build-system/tasks/report-test-status.js:121]: Thanks Raghu, I'm going that makes perfect sense to me. This essentially leads to a silent failure as is so it would be nice to log something.

File: build-system/tasks/visual-diff/index.js
Comment (rileyajones) [reply build-system/tasks/visual-diff/index.js]: Yeah, I reverted the whole file. I intend to fix types in the tasks/visual-diff directory all in one separate PR

**Review by Samerrr22 — APPROVED** [review]

**Review by Samerrr22 — APPROVED** [review]

File: build-system/pr-check/experiment-build.js
Comment (estherkim) [inline build-system/pr-check/experiment-build.js:35]: nit: remove `/** */` since there's no more comment

File: build-system/tasks/sweep-experiments/index.js
Comment (alanorozco) [reply build-system/tasks/sweep-experiments/index.js:262]: Sorry, I had this missed this question!

This seems fine per format received: `%h %aI %s`, though we have the conditional `.pop()` above. I'd say this is good but I'll make a separate change to put up a guard before this cast.

Comment (alanorozco) [reply build-system/tasks/sweep-experiments/index.js:262]: #33266 for safer parsing/testing

File: build-system/pr-check/experiment-build.js
Comment (rileyajones) [reply build-system/pr-check/experiment-build.js:35]: I plan to enable JSDoc enforcement soon to stem the amount of undocumented functions being added and removing these would slow that down. I could try to write an actual description of each of these functions or else I could add TODOs in each of these places, but both of those seem like pretty significant changes.

Comment (rsimha) [reply build-system/pr-check/experiment-build.js:35]: I think the main objection is in having a blank JSDoc. When you add the check, what will this end up being?

Also, is enforcement applicable only to those functions where there ought to be a non-empty JSDoc? Or will it insist on a blank `/** */`? (We should try and avoid this.)

Comment (rileyajones) [reply build-system/pr-check/experiment-build.js:35]: I definitely understand the objection but yes, it will insist on at least a blank comment. Given that there are 84 occurrences of this at them moment I would like to remove all these for now and try to add them all back in a better way in another dedicated PR.

File: build-system/tasks/sweep-experiments/index.js
Comment (rileyajones) [reply build-system/tasks/sweep-experiments/index.js:262]: Thanks @alanorozco

File: build-system/tasks/report-test-status.js
Comment (rsimha) [reply build-system/tasks/report-test-status.js:120]: With https://github.com/ampproject/amphtml/pull/33243/files#diff-25ae9c21c4de52ea749b2b460f0eba8459fcb2f07321d1428e4e5cf7a3406239R80, we're no longer returning `null`.

Comment (rileyajones) [reply build-system/tasks/report-test-status.js:120]: That's fantastic! I saw that PR but didn't realize it had landed.

**Review by estherkim — APPROVED** [review]
Thanks & sorry for the delay! I have a hard timing telling when something is ready for re-review.

**Pull request merged**
